core/interop/tests/class.rs

```diff
@@ -0,0 +1,112 @@
+//! Test for the class proc-macro.
+#![allow(unused_crate_dependencies)]
+
+use boa_engine::{js_string, Context, JsObject, JsString, Source};
+use boa_macros::{boa_class, Finalize, JsData, Trace};
+
+#[derive(Clone, Trace, Finalize, JsData)]
+enum AnimalType {
+    Cat,
+    Dog,
+    Other,
+}
+
+#[derive(Clone, Trace, Finalize, JsData)]
+struct Animal {
+    ty: AnimalType,
+    age: i32,
+}
+
+#[boa_class]
+#[boa(rename = "camelCase")]
+impl Animal {
+    #[boa(constructor)]
+    #[allow(clippy::needless_pass_by_value)]
+    fn new(name: String, age: i32) -> Self {
+        let ty = match name.as_str() {
+            "cat" => AnimalType::Cat,
+            "dog" => AnimalType::Dog,
+            _ => AnimalType::Other,
+        };
+
+        Self { ty, age }
+    }
+
+    fn static_method() -> i32 {
+        42
+    }
+
+    // Force this being a method (instead of a static function) by declaring it
+    // as a method.
+    #[boa(method)]
+    #[boa(length = 11)]
+    fn method(context: &mut Context) -> JsObject {
+        let obj = JsObject::with_null_proto();
+        obj.set(js_string!("key"), 43, false, context).unwrap();
+        obj
+    }
+
+    #[boa(getter)]
+    fn age(&self) -> i32 {
+        self.age
+    }
+
+    #[boa(setter)]
+    #[boa(method)]
```

Comment:
question: does this need to marked as method, it has `&self` so should it be enough to infer it?

Reply:
Yes, because it's marked as `setter`. It's a `method` by default if 2 conditions are met:

1. &self / &mut Self, and
2. it isn't marked `constructor`, `getter`, `setter` or `static`.